File: examples/go/sqs_test.go
package main

import (
	"context"
	"net"
	"net/http"
	"testing"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
)

const sqsPort = 10214

func createSQSClient(t *testing.T, port int, httpClient *http.Client) *sqs.Client {
	t.Helper()
	endpoint := endpointURL(port)
	cfg, err := config.LoadDefaultConfig(context.TODO(),
		config.WithRegion("ap-northeast-1"),
		config.WithCredentialsProvider(
			credentials.NewStaticCredentialsProvider("dummy", "dummy", ""),
		),
	)
	if err != nil {
		t.Fatalf("failed to load config: %v", err)
	}
	opts := []func(*sqs.Options){
		func(o *sqs.Options) {
			o.BaseEndpoint = aws.String(endpoint)
			o.RetryMaxAttempts = 1
		},
	}
	if httpClient != nil {
		opts = append(opts, func(o *sqs.Options) {
			o.HTTPClient = httpClient
		})
	}
	return sqs.NewFromConfig(cfg, opts...)
}

// --- Normal operation ---

func TestSQS_CreateQueue(t *testing.T) {
	client := createSQSClient(t, sqsPort, nil)
	result, err := client.CreateQueue(context.TODO(), &sqs.CreateQueueInput{
		QueueName: aws.String("go-test-queue"),
	})
	if err != nil {
		t.Fatalf("CreateQueue failed: %v", err)
	}
	if result.QueueUrl == nil {
		t.Fatal("QueueUrl is nil")
	}
	t.Logf("QueueUrl: %s", *result.QueueUrl)
}

func TestSQS_GetQueueUrl(t *testing.T) {
	client := createSQSClient(t, sqsPort, nil)
	result, err := client.GetQueueUrl(context.TODO(), &sqs.GetQueueUrlInput{
		QueueName: aws.String("go-test-queue"),
	})
	if err != nil {
		t.Fatalf("GetQueueUrl failed: %v", err)
	}
	if result.QueueUrl == nil {
		t.Fatal("QueueUrl is nil")
	}
}

func TestSQS_SendMessage(t *testing.T) {
	client := createSQSClient(t, sqsPort, nil)
	queueURL := endpointURL(sqsPort) + "/go-test-queue"
	result, err := client.SendMessage(context.TODO(), &sqs.SendMessageInput{
		QueueUrl:    aws.String(queueURL),
		MessageBody: aws.String("Hello from Go"),
	})
	if err != nil {
		t.Fatalf("SendMessage failed: %v", err)
	}
	if result.MessageId == nil {
		t.Fatal("MessageId is nil")
	}
	if result.MD5OfMessageBody == nil {
		t.Fatal("MD5OfMessageBody is nil")
	}
}

func TestSQS_ReceiveMessage(t *testing.T) {
	client := createSQSClient(t, sqsPort, nil)
	queueURL := endpointURL(sqsPort) + "/go-test-queue"
	result, err := client.ReceiveMessage(context.TODO(), &sqs.ReceiveMessageInput{
		QueueUrl: aws.String(queueURL),
	})
	if err != nil {
		t.Fatalf("ReceiveMessage failed: %v", err)
	}
	if len(result.Messages) == 0 {
		t.Fatal("no messages received")
	}
	msg := result.Messages[0]
	if msg.Body == nil {
		t.Fatal("message Body is nil")
	}
	if msg.MessageId == nil {
		t.Fatal("message MessageId is nil")
	}
	if msg.ReceiptHandle == nil {
		t.Fatal("message ReceiptHandle is nil")
	}
}

func TestSQS_DeleteMessage(t *testing.T) {
	client := createSQSClient(t, sqsPort, nil)
	queueURL := endpointURL(sqsPort) + "/go-test-queue"
	_, err := client.DeleteMessage(context.TODO(), &sqs.DeleteMessageInput{
		QueueUrl:      aws.String(queueURL),
		ReceiptHandle: aws.String("dummy-receipt-handle"),
	})
	if err != nil {
		t.Fatalf("DeleteMessage failed: %v", err)
	}
}

func TestSQS_DeleteQueue(t *testing.T) {
	client := createSQSClient(t, sqsPort, nil)
	queueURL := endpointURL(sqsPort) + "/go-test-queue"
	_, err := client.DeleteQueue(context.TODO(), &sqs.DeleteQueueInput{
		QueueUrl: aws.String(queueURL),
	})
	if err != nil {
		t.Fatalf("DeleteQueue failed: %v", err)
	}
}

// --- Fault injection: demonstrating the default timeout problem ---
//
// Go's http.DefaultClient has Timeout = 0 (no timeout).
// The AWS SDK for Go v2 also does not set a default timeout.
// This means requests to misbehaving servers will hang indefinitely.

func TestSQS_AcceptButSilent_DefaultHangs(t *testing.T) {
	// PITFALL: default client has no timeout — request hangs forever.
	// We use context.WithTimeout as a safety net to prove the SDK won't timeout on its own.
	client := createSQSClient(t, portAcceptButSilent, nil)
	start := time.Now()
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	_, err := client.CreateQueue(ctx, &sqs.CreateQueueInput{
		QueueName: aws.String("hang-test"),
	})
	elapsed := time.Since(start)
	if err == nil {
		t.Fatal("expected an error but call succeeded")
	}
	if elapsed < 2500*time.Millisecond {
		t.Fatalf("expected to hang for ~3s, but returned in %v", elapsed)
	}
	t.Logf("PITFALL confirmed: default client hung for %v until context cancelled", elapsed)
}

func TestSQS_AcceptButSilent_FixWithContextTimeout(t *testing.T) {
	// FIX: context.WithTimeout provides a request-scoped deadline.
	client := createSQSClient(t, portAcceptButSilent, nil)
	start := time.Now()
	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel()
	_, err := client.CreateQueue(ctx, &sqs.CreateQueueInput{
		QueueName: aws.String("timeout-test"),
	})
	elapsed := time.Since(start)
	if err == nil {
		t.Fatal("expected an error but call succeeded")
	}
	if elapsed >= 2500*time.Millisecond {
		t.Fatalf("expected timeout within ~1s, but took %v", elapsed)
	}
	t.Logf("FIX confirmed: context.WithTimeout cancelled in %v", elapsed)
}

func TestSQS_AcceptButSilent_FixWithHttpClientTimeout(t *testing.T) {
	// FIX: http.Client.Timeout covers the full request lifecycle.
	httpClient := &http.Client{Timeout: 1 * time.Second}
	client := createSQSClient(t, portAcceptButSilent, httpClient)
	start := time.Now()
	_, err := client.CreateQueue(context.TODO(), &sqs.CreateQueueInput{
		QueueName: aws.String("timeout-test"),
	})
	elapsed := time.Since(start)
	if err == nil {
		t.Fatal("expected an error but call succeeded")
	}
	if elapsed >= 2500*time.Millisecond {
		t.Fatalf("expected timeout within ~1s, but took %v", elapsed)
	}
	t.Logf("FIX confirmed: http.Client.Timeout cancelled in %v", elapsed)
}

func TestSQS_SlowResponse_DefaultHangs(t *testing.T) {
	// PITFALL: default client hangs on slow response — no timeout at all.
	client := createSQSClient(t, portSlowResponse, nil)
	start := time.Now()
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()
	_, err := client.CreateQueue(ctx, &sqs.CreateQueueInput{
		QueueName: aws.String("hang-test"),
	})
	elapsed := time.Since(start)
	if err == nil {
		t.Fatal("expected an error but call succeeded")
	}
	if elapsed < 2500*time.Millisecond {
		t.Fatalf("expected to hang for ~3s, but returned in %v", elapsed)
	}
	t.Logf("PITFALL confirmed: default client hung for %v on slow response", elapsed)
}

func TestSQS_SlowResponse_ResponseHeaderTimeoutDoesNotCoverBody(t *testing.T) {
	// PITFALL: ResponseHeaderTimeout only covers time to response headers.
	// SlowResponse returns HTTP 200 + headers immediately, then trickles body at 1 byte/sec.
	// ResponseHeaderTimeout is satisfied and the body read (during SDK deserialization) hangs.
	//
	// Unlike S3 GetObject where the body is returned as a stream to the caller,
	// SQS responses are fully deserialized by the SDK. So the hang occurs inside send().
	httpClient := &http.Client{
		Transport: &http.Transport{
			ResponseHeaderTimeout: 1 * time.Second,
			DialContext: (&net.Dialer{
				Timeout: 5 * time.Second,
			}).DialContext,
		},
	}
	client := createSQSClient(t, portSlowResponse, httpClient)
	start := time.Now()
	ctx, cancel := context.WithTimeout(context.Background(), 4*time.Second)
	defer cancel()
	_, err := client.CreateQueue(ctx, &sqs.CreateQueueInput{
		QueueName: aws.String("slow-body-test"),
	})
	elapsed := time.Since(start)
	if err == nil {
		t.Fatal("expected an error but call succeeded")
	}
	// ResponseHeaderTimeout didn't help — headers arrived in time.
	// Only the context deadline (4s) or another mechanism can save us.
	if elapsed < 2500*time.Millisecond {
		t.Fatalf("expected to hang beyond ResponseHeaderTimeout, but returned in %v", elapsed)
	}
	t.Logf("PITFALL confirmed: ResponseHeaderTimeout did not cover body read, hung for %v", elapsed)
}

func TestSQS_SlowResponse_FixWithHttpClientTimeout(t *testing.T) {
	// FIX: http.Client.Timeout covers the FULL lifecycle including body reads.
	// Unlike Node.js requestTimeout, Go's http.Client.Timeout also covers body consumption.
	httpClient := &http.Client{Timeout: 2 * time.Second}
	client := createSQSClient(t, portSlowResponse, httpClient)
	start := time.Now()
	_, err := client.CreateQueue(context.TODO(), &sqs.CreateQueueInput{
		QueueName: aws.String("timeout-test"),
	})
	elapsed := time.Since(start)
	if err == nil {
		t.Fatal("expected an error but call succeeded")
	}
	if elapsed >= 3*time.Second {
		t.Fatalf("expected timeout within ~2s, but took %v", elapsed)
	}
	t.Logf("FIX confirmed: http.Client.Timeout cancelled slow body read in %v", elapsed)
}

File: examples/php/tests/SqsTest.php
<?php

declare(strict_types=1);

namespace Rodriguez\Example\Tests;

use Aws\Sqs\SqsClient;
use PHPUnit\Framework\TestCase;

class SqsTest extends TestCase
{
    private const SQS_PORT = 10214;
    private const PORT_SLOW_RESPONSE = 10205;
    private const PORT_ACCEPT_BUT_SILENT = 10209;

    private static function createSqsClient(int $port = self::SQS_PORT, array $httpOptions = []): SqsClient
    {
        return new SqsClient([
            'region' => 'ap-northeast-1',
            'endpoint' => "http://localhost:{$port}",
            'credentials' => [
                'key' => 'dummy',
                'secret' => 'dummy',
            ],
            'retries' => 0,
            'http' => $httpOptions,
        ]);
    }

    // --- Normal operation ---

    public function testCreateQueue(): void
    {
        $sqs = self::createSqsClient();
        $result = $sqs->createQueue(['QueueName' => 'php-test-queue']);
        $this->assertNotEmpty($result['QueueUrl']);
    }

    public function testGetQueueUrl(): void
    {
        $sqs = self::createSqsClient();
        $result = $sqs->getQueueUrl(['QueueName' => 'php-test-queue']);
        $this->assertNotEmpty($result['QueueUrl']);
    }

    public function testSendMessage(): void
    {
        $sqs = self::createSqsClient();
        $queueUrl = "http://localhost:" . self::SQS_PORT . "/php-test-queue";
        $result = $sqs->sendMessage([
            'QueueUrl' => $queueUrl,
            'MessageBody' => 'Hello from PHP',
        ]);
        $this->assertNotEmpty($result['MessageId']);
        $this->assertNotEmpty($result['MD5OfMessageBody']);
    }

    public function testReceiveMessage(): void
    {
        $sqs = self::createSqsClient();
        $queueUrl = "http://localhost:" . self::SQS_PORT . "/php-test-queue";
        $result = $sqs->receiveMessage(['QueueUrl' => $queueUrl]);
        $this->assertNotEmpty($result['Messages']);
        $msg = $result['Messages'][0];
        $this->assertNotEmpty($msg['Body']);
        $this->assertNotEmpty($msg['MessageId']);
        $this->assertNotEmpty($msg['ReceiptHandle']);
    }

    public function testDeleteMessage(): void
    {
        $sqs = self::createSqsClient();
        $queueUrl = "http://localhost:" . self::SQS_PORT . "/php-test-queue";
        $result = $sqs->deleteMessage([
            'QueueUrl' => $queueUrl,
            'ReceiptHandle' => 'dummy-receipt-handle',
        ]);
        $this->assertLessThan(300, $result['@metadata']['statusCode']);
    }

    public function testDeleteQueue(): void
    {
        $sqs = self::createSqsClient();
        $queueUrl = "http://localhost:" . self::SQS_PORT . "/php-test-queue";
        $result = $sqs->deleteQueue(['QueueUrl' => $queueUrl]);
        $this->assertLessThan(300, $result['@metadata']['statusCode']);
    }

    // --- Fault injection: demonstrating the default timeout problem ---
    //
    // Guzzle's default `timeout` is 0 (no timeout).
    // The AWS SDK for PHP does not set a default timeout.
    // This means requests to misbehaving servers will hang indefinitely.
    //
    // Unlike Node.js SDK v3, Guzzle's `timeout` maps to CURLOPT_TIMEOUT which
    // covers the FULL transfer lifecycle including response body reads.
    // This means once `timeout` is configured, it protects against both
    // no-response and slow-body scenarios — a key advantage over Node.js.

    public function testAcceptButSilentDefaultHangs(): void
    {
        // PITFALL: default client has no timeout — request hangs forever.
        // We set a 3s timeout as safety net.
        $sqs = self::createSqsClient(self::PORT_ACCEPT_BUT_SILENT, [
            'timeout' => 3.0,
        ]);
        $start = microtime(true);
        try {
            $sqs->createQueue(['QueueName' => 'hang-test']);
            $this->fail('Expected exception');
        } catch (\Exception $e) {
            $elapsed = microtime(true) - $start;
            $this->assertGreaterThanOrEqual(2.5, $elapsed,
                "Expected to hang for ~3s, but returned in {$elapsed}s");
            $this->addToAssertionCount(1);
        }
    }

    public function testAcceptButSilentFixWithTimeout(): void
    {
        // FIX: Guzzle timeout covers the entire transfer.
        $sqs = self::createSqsClient(self::PORT_ACCEPT_BUT_SILENT, [
            'timeout' => 1.0,
        ]);
        $start = microtime(true);
        try {
            $sqs->createQueue(['QueueName' => 'timeout-test']);
            $this->fail('Expected exception');
        } catch (\Exception $e) {
            $elapsed = microtime(true) - $start;
            $this->assertLessThan(2.5, $elapsed,
                "Expected timeout within ~1s, but took {$elapsed}s");
            $this->assertMatchesRegularExpression('/timed?\s*out/i', $e->getMessage());
        }
    }

    public function testSlowResponseDefaultHangs(): void
    {
        // PITFALL: default client hangs on slow response — no timeout.
        $sqs = self::createSqsClient(self::PORT_SLOW_RESPONSE, [
            'timeout' => 3.0,
        ]);
        $start = microtime(true);
        try {
            $sqs->createQueue(['QueueName' => 'hang-test']);
            $this->fail('Expected exception');
        } catch (\Exception $e) {
            $elapsed = microtime(true) - $start;
            $this->assertGreaterThanOrEqual(2.5, $elapsed,
                "Expected to hang for ~3s, but returned in {$elapsed}s");
            $this->addToAssertionCount(1);
        }
    }

    public function testSlowResponseFixWithTimeout(): void
    {
        // FIX: Guzzle timeout (CURLOPT_TIMEOUT) covers the FULL lifecycle including body reads.
        // Unlike Node.js requestTimeout, this works for slow body reads too.
        $sqs = self::createSqsClient(self::PORT_SLOW_RESPONSE, [
            'timeout' => 2.0,
        ]);
        $start = microtime(true);
        try {
            $sqs->createQueue(['QueueName' => 'timeout-test']);
            $this->fail('Expected exception');
        } catch (\Exception $e) {
            $elapsed = microtime(true) - $start;
            $this->assertLessThan(3.0, $elapsed,
                "Expected timeout within ~2s, but took {$elapsed}s");
        }
    }
}
